### PR TITLE
Add utility functions: (1) plotting ROC curve and (2) function to calculate Kagan_I_1_score

### PR DESCRIPTION
**Pull request by hanbao-ucla**

### CODE REVIEW

**Comment by wsavran** [comment]
right on! ill take a look at this in some more detail on monday. just so you know, you can keep posting commits to this pull request if you'd like to make changes. no need to close and re-open.

**Comment by hanbao-ucla** [comment]
Oh! Didn't know that! Do I just make changed in my own branch? I mean the branch that I'm trying to push to the upstream master?

**Comment by wsavran** [comment]
Yes, every commit that you make to the branch `csep-feat-han` will show up in this pull request. it will re-build and test to make sure that any new commits don't break the build. its a cool feature.

**Comment by hanbao-ucla** [comment]
Sounds great. Thanks Bill! 
Btw, I've tested against the most up-to-date upstream master version and it works well.

**Comment by wsavran** [comment]
great ill post in this thread as soon as i look over everything!

**Comment by hanbao-ucla** [comment]
No rush, Bill! ;)

**Comment by wsavran** [comment]
hi @hanbao-ucla, sorry its taken me so long to look at this, but im ready to start moving toward getting these new functions into pycsep. im going to put some code review notes in there. specifically, there are some utility functions that exist that should be used and we want to make the overall functionality of the plotting routines similiar. eg., in tthe other functions we accept multiple forecasts and a single testing catalog. im going to put some mroe notes into the code review.

**Comment by hanbao-ucla** [comment]
Hi Bill,

No problem. Feel free to let me know if there is anything you need me to do. :)

**Comment by wsavran** [comment]
i left a few comments in the code review. let me know what you think about them

**Comment by hanbao-ucla** [comment]
Bill, I don't see your code review comment.

**Review by wsavran — CHANGES_REQUESTED** [review]

**Comment by wsavran** [comment]
my apologies! it looks like i never push the final 'submit' button

File: csep/utils/plots.py [hunk 1]
```diff
@@ -1442,4 +1442,99 @@ def plot_calibration_test(evaluation_result, axes=None, plot_args=None, show=Fal
...
+            TESTED EVENT FALLS IN)
+    """
+    ### Determine if input 'catalogs' is a list of csep.catalogs or a single csep.catalog
+    try:
+        N_cat = len(catalogs) # the input catalogs is a list of csep.catalogs
+    except:  
+        N_cat = 1             # the input catalogs is a single csep.catalog
+        catalogs = [catalogs]
+    
+    earth_radius_km = 6371.
```
[the remark this answers is not in the document]
Comment (hanbao-ucla) [reply csep/utils/plots.py:1473]: Agree.

File: csep/utils/plots.py [hunk 2]
```diff
@@ -1442,4 +1442,99 @@ def plot_calibration_test(evaluation_result, axes=None, plot_args=None, show=Fal
...
+        catalogs:  csep.catalog or a list of csep.catalog (catalogs to test against the forecast)
+        
+    Note that: 
+        (1) the testing catalog and forecast should have exactly the same time-window (duration)
+        (2) grids of forecast has to be regular rectangular shape that are defined by 
+            [min_lat,max_lat,d_lat; min_lon,max_lon,d_lon]. And the grids are ordered as such:
+            it scan through the rectangular region from [min_lat,min_lon] to [max_lat,min_lon], 
+            than move to next longitude line, i.e. from [min_lat+d_lat,min_lon] to [max_lat+d_lat,min_lon].
+            The gridded region can be the entire globe. (THIS IS IMPORTANT TO FIND WHICH BIN DOES EACH
+            TESTED EVENT FALLS IN)
```
[the remark this answers is not in the document]
Comment (hanbao-ucla) [reply csep/utils/plots.py:1464]: Got it!

File: csep/utils/plots.py [hunk 3]
```diff
@@ -1442,4 +1442,99 @@ def plot_calibration_test(evaluation_result, axes=None, plot_args=None, show=Fal
...
+    d_lon = lons[1] - lons[0]                      # get grid interval [d_lon]
+    
+    area_km2 = np.zeros(bin_lon.shape, dtype=np.float64) # Initialze
+    for i, bot_lat in enumerate(bin_lat):
+        bot_lon = bin_lon[i]
+        top_lat = bot_lat + d_lat
+        top_lon = bot_lon + d_lon
+        strip_area_steradian = 2 * np.pi * (1.0e0 - np.cos((90.0e0 - bot_lat) * rad_per_deg)) \
+                             - 2 * np.pi * (1.0e0 - np.cos((90.0e0 - top_lat) * rad_per_deg))
+        area_km2[i] = strip_area_steradian * R2 / (360.0/d_lon)
```
[the remark this answers is not in the document]
Comment (hanbao-ucla) [reply csep/utils/plots.py:1495]: Sounds great to me! If so, we will have the dimension array of 'Area' same as that of the 'Bin-lat' or 'Bin-lon'

File: csep/utils/plots.py [hunk 4]
```diff
@@ -1442,4 +1442,99 @@ def plot_calibration_test(evaluation_result, axes=None, plot_args=None, show=Fal
...
+    # Get normalized & sorted [prob] (probability) and [area]
+    rate = forecast.spatial_counts() # get [rate per cell] for all cells
```
[the remark this answers is not in the document]
Comment (hanbao-ucla) [reply csep/utils/plots.py:1497]: Thanks! Yeah you are right.

File: csep/utils/plots.py [hunk 5]
```diff
@@ -1442,4 +1442,99 @@ def plot_calibration_test(evaluation_result, axes=None, plot_args=None, show=Fal
...
+    I = np.argsort(rate)             # get index of descending sort
```
[the remark this answers is not in the document]
Comment (hanbao-ucla) [reply csep/utils/plots.py:1498]: Oh, I see. Yeah... I was commenting the two lines (line 1498 & 1499) together. But you are right.

File: csep/utils/plots.py [hunk 6]
```diff
@@ -1442,4 +1442,99 @@ def plot_calibration_test(evaluation_result, axes=None, plot_args=None, show=Fal
...
+    I = np.flip(I)                   # get index of descending sort
+    area_norm_sorted = np.cumsum(area_km2[I])/ np.sum(area_km2)
+    prob_norm_sorted = np.cumsum(rate[I])    / np.sum(rate)
+    
+    ### GET information score contribution from each testing catalog
+    cats_norm_sorted   = np.zeros((N_cat,len(rate)), dtype=np.float64) # initialze 
+    for j,subcatalog in enumerate(catalogs):
```
[the remark this answers is not in the document]
Comment (hanbao-ucla) [reply csep/utils/plots.py:1505]: Oh! Haha, this is just what I was doing. I was testing one forecast against different years and different magnitudes.... If this is abnormal, we can definitely change it.

File: csep/utils/plots.py [hunk 7]
```diff
@@ -1442,4 +1442,99 @@ def plot_calibration_test(evaluation_result, axes=None, plot_args=None, show=Fal
...
+        
+        evt_counts = np.zeros(rate.shape, dtype=np.int32) # Initialize (number of events in all bins)
+        for i in np.arange(N_event):
+            idx = int( np.floor((evt_lon[i] - (min_lon)) * (1.0/d_lon)) ) * len(lats) \
+                + int( np.floor((evt_lat[i] - (min_lat)) * (1.0/d_lat)) ) 
+            evt_counts[idx] = evt_counts[idx] + 1
+        cats_norm_sorted[j,:] = np.cumsum(evt_counts[I])/np.sum(evt_counts)
+    
+    ### PLOT ROC Curves
+    # outFile = 'ROC_comp_M5M7.png'
```
[the remark this answers is not in the document]
Comment (hanbao-ucla) [reply csep/utils/plots.py:1518]: Thanks for picking that up!

File: csep/utils/plots.py [hunk 8]
```diff
@@ -1442,4 +1442,99 @@ def plot_calibration_test(evaluation_result, axes=None, plot_args=None, show=Fal
...
+    for j,subcatalog in enumerate(catalogs):
+        evt_lat = subcatalog.get_latitudes()    # event location in the testing catalog
+        evt_lon = subcatalog.get_longitudes()   # event location in the testing catalog
+        N_event = subcatalog.event_count        # total number of events of the testing catalog
+        
+        evt_counts = np.zeros(rate.shape, dtype=np.int32) # Initialize (number of events in all bins)
+        for i in np.arange(N_event):
+            idx = int( np.floor((evt_lon[i] - (min_lon)) * (1.0/d_lon)) ) * len(lats) \
+                + int( np.floor((evt_lat[i] - (min_lat)) * (1.0/d_lat)) ) 
+            evt_counts[idx] = evt_counts[idx] + 1
```
[the remark this answers is not in the document]
Comment (hanbao-ucla) [reply csep/utils/plots.py:1514]: Thanks! Yeah it make sense to me.

File: csep/utils/plots.py [hunk 9]
```diff
@@ -1442,4 +1442,99 @@ def plot_calibration_test(evaluation_result, axes=None, plot_args=None, show=Fal
...
+        name='ROC Curves'
+    fig.suptitle(name,fontsize=18)
+    
+    if savepdf:
+        outFile = "{}.pdf".format(name)
+        plt.savefig(outFile,format='pdf')
+    if savepng:
+        outFile = "{}.png".format(name)
+        plt.savefig(outFile,format='png')
+    return fig, ax
```
[the remark this answers is not in the document]
Comment (hanbao-ucla) [reply csep/utils/plots.py:1540]: Message well received!

**Comment by hanbao-ucla** [comment]
Thanks Bill. You comments are right on.

File: csep/utils/plots.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (khawajasim) [reply csep/utils/plots.py:1495]: that would be great.

**Comment by wsavran** [comment]
@hanbao-ucla i just merged a pull request that adds the area function into regions.py. you can just pull those changes into this pull request and use that area function.

File: csep/utils/plots.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hanbao-ucla) [reply csep/utils/plots.py:1515]: fixed.

File: csep/utils/plots.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (hanbao-ucla) [reply csep/utils/plots.py:1524]: Added in csep/utils/constants

File: csep/utils/plots.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hanbao-ucla) [reply csep/utils/plots.py:1548]: Fixed

File: csep/utils/plots.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (hanbao-ucla) [reply csep/utils/plots.py:1549]: fixed.

File: csep/utils/plots.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (hanbao-ucla) [reply csep/utils/plots.py:1569]: fixed.

File: csep/utils/plots.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (hanbao-ucla) [reply csep/utils/plots.py:1546]: fixed. Using the newly added function (by Bill Savran) core.regions.geographical_area_from_bounds(lon1,lat1,lon2,lat2)

File: csep/utils/stats.py [hunk 10]
```diff
@@ -205,3 +205,86 @@ def poisson_inverse_cdf(random_matrix, lam):
...
+       I_1
+    """
+    ### Determine if input 'catalogs' is a list of csep.catalogs or a single csep.catalog
+    try:
+        N_cat = len(catalogs) # the input catalogs is a list of csep.catalogs
+    except:  
+        N_cat = 1             # the input catalogs is a single csep.catalog
+        catalogs = [catalogs]
+            
+    earth_radius_km = 6371.
```
[the remark this answers is not in the document]
Comment (hanbao-ucla) [reply csep/utils/stats.py:247]: Fixed.

File: csep/utils/stats.py [hunk 11]
```diff
@@ -205,3 +205,86 @@ def poisson_inverse_cdf(random_matrix, lam):
...
+    d_lon = lons[1] - lons[0]                      # get grid interval [d_lon]
+    area_km2 = np.zeros(bin_lon.shape, dtype=np.float64) # Initialze
+    for i, bot_lat in enumerate(bin_lat):
+        bot_lon = bin_lon[i]
+        top_lat = bot_lat + d_lat
+        top_lon = bot_lon + d_lon
+        strip_area_steradian = 2 * np.pi * (1.0e0 - np.cos((90.0e0 - bot_lat) * rad_per_deg)) \
+                             - 2 * np.pi * (1.0e0 - np.cos((90.0e0 - top_lat) * rad_per_deg))
+        area_km2[i] = strip_area_steradian * R2 / (360.0/d_lon)
+    total_area = np.sum(area_km2) # Total Area
```
[the remark this answers is not in the document]
Comment (hanbao-ucla) [reply csep/utils/stats.py:269]: Fixed.  Using the newly added function (by Bill Savran) core.regions.geographical_area_from_bounds(lon1,lat1,lon2,lat2)

File: csep/utils/plots.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (hanbao-ucla) [reply csep/utils/plots.py:1565]: Fixed. Just need to feed the forecast.region to the catalog to using catalog.spatial_counts().
The code is much more beautiful now than before :)

File: csep/utils/stats.py [hunk 12]
```diff
@@ -205,3 +205,86 @@ def poisson_inverse_cdf(random_matrix, lam):
...
+        evt_lon = subcatalog.get_longitudes()    # event location in the testing catalog
+        N_event = subcatalog.event_count         # total number of events of the testing catalog
+
+        score    = np.zeros(evt_lon.shape, dtype=np.float64) # initialize
+        for i in np.arange(N_event):
+            idx = int( np.floor((evt_lon[i] - (min_lon)) * (1.0/d_lon)) ) * len(lats) \
+                + int( np.floor((evt_lat[i] - (min_lat)) * (1.0/d_lat)) ) 
+            score[i] = np.log2(rateDen[idx]/uniform_forecast)
+        ### GET I_1 score
+        I_1[j] = sum(score)/N_event
```
[the remark this answers is not in the document]
Comment (hanbao-ucla) [reply csep/utils/stats.py:288]: Fixed. Now using function catalog.spatial_counts()

File: csep/utils/plots.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (hanbao-ucla) [reply csep/utils/plots.py:1556]: Fixed. Now accepting single or multiple forecasts but only one catalog.

File: csep/utils/plots.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (hanbao-ucla) [reply csep/utils/plots.py:1591]: Fixed the axes input/output problem as your suggestion.

I don't think the current ROC curve has any EvaluationResult (values for further evaluation).

**Comment by hanbao-ucla** [comment]
Hi Bill,

I've addressed all of your comments, and I think it is now ready to be merged. :)

++Han

**Comment by wsavran** [comment]
thats great @hanbao-ucla ! i'm running the automated build and test now. ill review the code and hopefully merge tomorrow!

**Comment by hanbao-ucla** [comment]
All bugs fixed. Documentations updated. I think it can be merged now.

**Comment by wsavran** [comment]
thanks for the updates @hanbao-ucla. now that the ci tools are fixed im going to checkout this pull request.

**Pull request merged**